src/python/pants/backend/debian/rules.py

```diff
@@ -39,55 +43,58 @@ class DebianPackageFieldSet(PackageFieldSet):
             binary_name="dpkg-deb",
             search_path=["/usr/bin"],
         ),
+        **implicitly(),
     )
     if not dpkg_deb_path.first_path:
         raise OSError(f"Could not find the `{dpkg_deb_path.binary_name}` program in `/usr/bin`.")
 
-    hydrated_sources = await Get(HydratedSources, HydrateSourcesRequest(field_set.sources_dir))
+    hydrated_sources = await hydrate_sources(
+        HydrateSourcesRequest(field_set.sources_dir), **implicitly()
+    )
 
     # Since all the sources are coming only from a single directory, it is
     # safe to pick an arbitrary file and get its root directory name.
     # Validation of the resolved files has been called on the target, so it is known that
     # snapshot.files isn't empty.
     sources_directory_name = PurePath(hydrated_sources.snapshot.files[0]).parts[0]
 
-    result = await Get(
-        ProcessResult,
-        Process(
-            argv=(
-                dpkg_deb_path.first_path.path,
-                "--build",
-                sources_directory_name,
-            ),
-            description="Create a Debian package from the produced packages.",
-            input_digest=hydrated_sources.snapshot.digest,
-            # dpkg-deb produces a file with the same name as the input directory
-            output_files=(f"{sources_directory_name}.deb",),
-            env={"PATH": str(PurePath(tar_binary_path.path).parent)},
-        ),
+    result = await fallible_to_exec_result_or_raise(
+        **implicitly(
+            Process(
+                argv=(
+                    dpkg_deb_path.first_path.path,
+                    "--build",
+                    sources_directory_name,
+                ),
+                description="Create a Debian package from the produced packages.",
+                input_digest=hydrated_sources.snapshot.digest,
+                # dpkg-deb produces a file with the same name as the input directory
+                output_files=(f"{sources_directory_name}.deb",),
+                env={"PATH": str(PurePath(tar_binary_path.path).parent)},
+            )
+        )
     )
     # The output Debian package file needs to be renamed to match the output_path field.
     output_filename = field_set.output_path.value_or_default(
         file_ending="deb",
     )
-    digest_entries = await Get(DigestEntries, Digest, result.output_digest)
+    digest_entries = await directory_digest_to_digest_entries(result.output_digest)
     assert len(digest_entries) == 1
     result_file_entry = digest_entries[0]
     assert isinstance(result_file_entry, FileEntry)
     new_file = FileEntry(output_filename, result_file_entry.file_digest)
 
-    final_result = await Get(Digest, CreateDigest([new_file]))
+    final_result = await create_digest_to_digest(CreateDigest([new_file]))
     return BuiltPackage(final_result, artifacts=(BuiltPackageArtifact(output_filename),))
 
 
-def rules():
-    return [
+def rules() -> Iterable[Rule | UnionRule]:
```

Comment:
Should we consider exporting a `Rules` type, making this more ergonomic, and also makes it easier in the future to add/adjust this interface with less code impact?
```python
from pants.engine.rules import Rules

def rules() -> Rules:
  return ( ... )

# pants.engine.rules

Rules = TypeAlias[Iterable[Rule | UnionRule]]
```

Reply:
In general, yes - but I think that should be a separate PR ... and done with a linter. I realized how easy this would be to make automated after responding to this PR 😆

Reply:
#21125